File: mobile/src/api/evolution.ts
import { api } from "./client";

export type WeightPoint = { date: string; weight_kg: number };
export type VolumePoint = { date: string; volume_kg: number; sets: number };
export type ExerciseOption = { id: number; name: string; set_count: number };
export type ExerciseProgressionPoint = { date: string; max_weight_kg: number; volume_kg: number };

export async function getWeightEvolution(): Promise<WeightPoint[]> {
  const { data } = await api.get<WeightPoint[]>("/evolution/weight");
  return data;
}

export async function getVolumeEvolution(): Promise<VolumePoint[]> {
  const { data } = await api.get<VolumePoint[]>("/evolution/volume");
  return data;
}

export async function getExercisesWithHistory(): Promise<ExerciseOption[]> {
  const { data } = await api.get<ExerciseOption[]>("/evolution/exercises");
  return data;
}

export async function getExerciseProgression(
  exerciseId: number
): Promise<{ exercise_name: string; points: ExerciseProgressionPoint[] }> {
  const { data } = await api.get(`/evolution/exercise/${exerciseId}`);
  return data;
}

export type NutritionDay = { date: string; kcal: number };
export type NutritionHistory = {
  days: NutritionDay[];
  goal_kcal: number | null;
  days_logged: number;
  days_within_goal: number;
};

export async function getNutritionHistory(days = 14): Promise<NutritionHistory> {
  const { data } = await api.get<NutritionHistory>("/evolution/nutrition", { params: { days } });
  return data;
}

export type ConsistencyDay = {
  date: string;
  trained: boolean;
  slept_well: boolean;
  hydrated: boolean;
  logged_food: boolean;
  score: number;
};
export type ConsistencyHistory = {
  days: ConsistencyDay[];
  current_streak: number;
  best_streak: number;
};

export async function getConsistency(days = 30): Promise<ConsistencyHistory> {
  const { data } = await api.get<ConsistencyHistory>("/evolution/consistency", { params: { days } });
  return data;
}

File: backend/app/schemas/evolution.py
from datetime import datetime

from pydantic import BaseModel


class WeightPoint(BaseModel):
    date: datetime
    weight_kg: float


class VolumePoint(BaseModel):
    date: datetime
    volume_kg: float
    sets: int


class ExerciseOption(BaseModel):
    id: int
    name: str
    set_count: int


class ExerciseProgressionPoint(BaseModel):
    date: datetime
    max_weight_kg: float
    volume_kg: float


class ExerciseProgressionResponse(BaseModel):
    exercise_name: str
    points: list[ExerciseProgressionPoint]


class NutritionDay(BaseModel):
    date: str
    kcal: int
    protein_g: float = 0
    carbs_g: float = 0
    fat_g: float = 0


class NutritionHistoryResponse(BaseModel):
    days: list[NutritionDay]
    goal_kcal: float | None
    goal_protein_g: float | None = None
    goal_carbs_g: float | None = None
    goal_fat_g: float | None = None
    days_logged: int
    days_within_goal: int


class StrengthGroup(BaseModel):
    group: str  # "superiores" | "inferiores" | "core"
    avg_pct_change: float
    exercises_count: int


class StrengthByGroupResponse(BaseModel):
    groups: list[StrengthGroup]


class ConsistencyDay(BaseModel):
    date: str
    trained: bool
    slept_well: bool
    hydrated: bool
    logged_food: bool
    score: int


class ConsistencyResponse(BaseModel):
    days: list[ConsistencyDay]
    current_streak: int
    best_streak: int
